File: mlproject/Board.java
import java.util.Random;
import java.util.Arrays;

public class Board{

	private int[][] board;
	private int lastJ,lastI,n;
	private boolean turn;
	Board(int n){
		this.n = n;
		board = new int[n][n];
		turn = new Random().nextBoolean();
	}
	int getSize(){
		return n;
	}
	boolean getTurn(){
		return turn;
	}
	boolean isTurn(int mark){
		return (turn&&mark==Mark.x||mark==Mark.o&&!turn);
	}
	void clear(){
		turn = new Random().nextBoolean();
		board = new int[n][n];
	}
	char nextMove(){
		if(turn)
			return Mark.x;
		else
			return Mark.o;
	}
	boolean play(int i,int j){
		if(board[i][j]!=Mark.blank)
			return false;
		board[i][j] = nextMove();
		lastI = i;
		lastJ = j;
		turn = !turn;
		return true;
	}

	void undo(){
		board[lastI][lastJ] = 0;
		turn = !turn;
	}
	boolean isBlank(int i,int j){
		return board[i][j] == Mark.blank;

	}
	boolean isBlank(Position pos){
		return board[pos.i][pos.j] == Mark.blank;
	}
	boolean isGameOver(){
		if(hasWon())
			return true;
		for(int[] arr:board)
			for(int b:arr)
				if(b==Mark.blank)
					return false;
		return true;
	}
	boolean willWin(Position action){
		if(!isBlank(action))
			return false;
		int mark = nextMove();
		board[action.i][action.j] = mark;
		boolean willWin = false;
		if(hasWon(mark))
			willWin = true;
		board[action.i][action.j] = Mark.blank;
		return willWin;
	}
	boolean hasWon(){
		int mark = nextMove();
		mark = Mark.getOpposite(mark);
		return hasWon(mark);
	}
	boolean hasWon(int mark){
		for(int i=0;i<n;i++){
			int j;
			for(j=0;j<n;j++)
				if(board[i][j]!=mark)
					break;
			if(j>=n)
				return true;
		}
		for(int i=0;i<n;i++){
			int j;
			for(j=0;j<n;j++)
				if(board[j][i]!=mark)
					break;
			if(j>=n)
				return true;
		}
		int i;
		for(i=0;i<n;i++)
			if(board[i][i]!=mark)
				break;
		if(i>=n)
			return true;
		for(i=0;i<n;i++)
			if(board[i][n-i-1]!=mark)
				break;
		if(i>=n)
			return true;
		return false;
	}
	void print(){
		for(int[] arr : board){
			System.out.print("|");
			for(int a:arr){
				char b;
				if(a==Mark.blank)
					b = ' ';
				else if(a==Mark.x)
					b = 'X';
				else
					b = 'O';
				System.out.print(" "+b+" |");
			}
			System.out.print("\n-");
			for(int i=0;i<n;i++)
				System.out.print("----");
			System.out.println();
		}
	}
	int[] getState(){
		int[] state = new int[n*n];
		for(int i=0;i<n;i++)
			for(int j=0;j<n;j++){
				state[i*n+j] = board[i][j];
			}
		return state;
	}
}

File: python/board.py
#!/usr/bin/python
from __future__ import print_function
from mark import Mark
import random
class Board(object):
	"""docstring for Board"""
	def __init__(self, n):
		super(Board, self).__init__()
		self.n = n
		self.lastI=0
		self.lastJ=0
		board=[[0 for i in range(0,self.n)] for j in range(0,self.n)]
		turn=(random.random()<0.5)
	def getSize(self):
		return self.n
	def getTurn(self):
		return self.turn
	def isTurn(self,mark):
		return (self.turn and mark==Mark.x or mark==Mark.o and not self.turn)
	def clear(self):
		self.board=[[0 for i in range(0,self.n)] for j in range(0,self.n)]
		self.turn=(random.random()<0.5)
	def nextMove(self):
		if(self.turn):
			return Mark.x
		else:
			return Mark.o
	def play(self,i,j):
		if(self.board[i][j]!=Mark.blank):
			return False
		self.board[i][j]=self.nextMove()
		self.lastI=i
		self.lastJ=j
		self.turn=not self.turn
		return True
	def undo(self):
		self.board[self.lastI][self.lastJ]=0
		self.turn=not self.turn
	def isBlank(self,pos):
		return self.board[pos.i][pos.j]==Mark.blank
	def isGameOver(self):
		if(self.hasWon()):
			return True
		for arr in self.board:
			for b in arr:
				if(b==Mark.blank):
					return False
		return True
	def willWin(self,action):
		if not self.isBlank(action):
			return False
		mark=self.nextMove()
		self.board[action.i][action.j]=mark
		willWin=False
		if(self.hasWon(mark)):
			willWin=True
		self.board[action.i][action.j]=Mark.blank
		return willWin
	def hasWon(self,mark=-1):
		if mark==-1:
			mark=Mark.getOpposite(self.nextMove())
		for arr in self.board:
			if(arr==[mark for i in range(0,self.n)]):
				return True
		for i  in range(0,self.n):
			arr=[row[i] for row in self.board]
			if(arr==[mark for i in range(0,self.n)]):
				return True
		arr=[self.board[i][i] for i in range(0,self.n)]
		if(arr==[mark for i in range(0,self.n)]):
			return True
		arr=[self.board[i][self.n-i-1] for i in range(0,self.n)]
		if(arr==[mark for i in range(0,self.n)]):
			return True
		return False
	def prnt(self):
		for arr in self.board:
			print("|",end="")
			for a in arr:
				b=''
				if(a==Mark.blank):
					b=' '
				elif(a==Mark.x):
					b='X'
				else:
					b='O'
				print(" "+b+" |",end="")
			print("")
			print("-",end="")
			for i in range(0,self.n):
				print("----",end="")
			print("")
	def getState(self):
		state=[0 for i in range(0,self.n*self.n)]
		for i  in range(0,self.n):
			for j in range(0,self.n):
				state[i*self.n+j]
		return state

